### ISSUE TITLE
Reactions to messages

### ISSUE THREAD

**Issue by gabriel-vasile** [tinode/chat#693]
In frontends each message will have a "Add reaction" button which will open a pop-up with thumbs ups, thumbs downs, and other smiley faces and reactions. Users can choose a reaction which will be added to the list of reactions for that message.
Is this a feature you're willing to have in tinode, @or-else?

**Comment by or-else** [comment]
Is this the same/related to https://github.com/tinode/chat/issues/672?

**Comment by or-else** [comment]
Also this https://github.com/tinode/webapp/issues/49 and this https://github.com/tinode/webapp/issues/30

**Comment by or-else** [comment]
Is it for channels?

**Comment by gabriel-vasile** [comment]
No, it's not related to any of the 3 issues. Those ask for the messages to contain emojis, stickers, etc.
This is about adding reactions to already sent messages. Similar to what github, slack, messenger have.

> Is it for channels?

I was thinking only users with write permission to topic to be able to add reactions. So it would work on channels for subscribers, but readers would not be allowed to send reactions, they would just see them.

**Comment by rkgarcia** [comment]
What's the purpose with this? I can't see that options in products like whatsapp and telegram

**Comment by or-else** [comment]
We have discussed a similar feature internally. I created an issue for it for reference: https://github.com/tinode/chat/issues/696

Your feature is effectively a different UI for the polling. The reactions feature can be seen as an always-present poll with predefined options.

It's not a simple feature though. The main complexity is how to store the votes (reactions) on the backend and how to report them back. We wanted to use [{note}](https://github.com/tinode/chat/blob/next/docs/API.md#note) with what=data and a json payload to collect votes, use {info} to update stats while the user is attached to topic. But then we need to be able to enforce voting rules on the backend, aggregate statistics. It's not clear if the votes should be stored in a separate table/collection or in the messages table. Both have drawbacks.

**Comment by or-else** [comment]
> What's the purpose with this? I can't see that options in products like whatsapp and telegram

It's a useful feature for channels. FB/IG messenger has it. Telegram has a similar but heavier feature: channels can be enabled to have a conversation for each message.

**Comment by gabriel-vasile** [comment]
> Your feature is effectively a different UI for the polling. The reactions feature can be seen as an always-present poll with predefined options.

No. Polls are linked to a conversation, reactions are linked to a message.
> #696 This feature should work in normals groups and in channels.

No. Reactions should work in any conversation: p2p, group, channel.

Again, what I said in the original post is: 
> EACH message will have a "Add reaction" button
> Users can choose a reaction which will be added to the list of reactions for THAT MESSAGE

Examples:
Reactions in skype:
![reactions skype](https://user-images.githubusercontent.com/10602896/130202368-bfc955e9-3709-405e-9e5c-50e6615b94bf.png)
Reactions in slack:
![reactions slack1](https://user-images.githubusercontent.com/10602896/130202434-a25f7e04-273a-4c1e-baaa-7531eb84cec5.png)
![reactions slack2](https://user-images.githubusercontent.com/10602896/130202444-71be0abc-40b8-49da-889d-263d262d4f69.png)

**Comment by or-else** [comment]
> No.
> No.

In such a case I have further input to share with you.

**Comment by vladdnepr** [comment]
Hi. Do you have any updates about this feature? Can I help with this one?

**Comment by or-else** [comment]
> Can I help with this one?

Sure but it's a complex feature. Particularly on the backed: how to store reactions, how to query for them, send updates. Your other feature request `Pinned messages, #751` is much simpler and maybe a better start if you wish to contribute.

**Comment by or-else** [comment]
The feature is in v0.26.0-alpha1
